std/encoding.d

```diff
@@ -3907,3 +3907,121 @@ if (isForwardRange!Range && is(Unqual!(ElementType!Range) == ubyte))
 
 /** Constant defining a fully decoded BOM */
 enum dchar utfBOM = 0xfeff;
+
+
+/++
+    Strips the BOM (if present) from the front of the given range and returns
+    the range.
+
+    If the range is a range of $(D ubyte), then $(LREF getBOM) will be used to
+    determined if there is a BOM present, and the matching BOM will be stripped.
+
+    If the range is a range of $(D char) or an array of $(D char), then the
+    UTF-8 BOM will be stripped if present, but no other BOMs will be checked
+    for.
+
+    If the range is a range of $(D wchar) or an array of $(D wchar), then the
+    UTF-16 BOM of the native endianness will be stripped if present, but no
+    other BOMs will be checked for.
+
+    If the range is a range of $(D dchar) or an array of $(D dchar), then the
+    UTF-32 BOM of the native endianness will be stripped if present, but no
+    other BOMs will be checked for.
+
+    Params:
+        range = A range of characters of ubyte.
+
+    Returns:
+        The range with the BOM stripped from the front of the given range if
+        there is a BOM at the front of the range; otherwise, the given range is
+        returned unchanged.
+  +/
+R stripBOM(R)(R range)
+if (isForwardRange!R && (isSomeChar!(ElementType!R) || is(Unqual!(ElementType!R) == ubyte)))
+{
+    import std.encoding : getBOM;
+
+    alias E = Unqual!(ElementEncodingType!R);
+
+    static if (is(E == ubyte))
+    {
+        range.popFrontN(range.save.getBOM().sequence.length);
+        return range;
+    }
+    else
+    {
+        import std.algorithm.searching : startsWith;
+        import std.utf : byCodeUnit;
+
+        auto bcu = range.byCodeUnit();
+
+        static if (is(E == char))
+        {
+            import std.range : only;
+            if (bcu.save.startsWith(only(cast(char)0xEF, cast(char)0xBB, cast(char)0xBF)))
+                bcu.popFrontN(3);
+        }
+        else static if (is(E == wchar))
+        {
+            if (bcu.save.startsWith(cast(wchar)0xFEFF))
+                bcu.popFront();
+        }
+        else
+        {
+            if (range.save.startsWith(cast(dchar)0x0000FEFF))
+                bcu.popFront();
+        }
```

Comment:
Reposting obsolete comment:

Any reason this entire `static if` couldn't be replaced with `if (range.front == '\uFEFF') range.popFront();` ?

Reply:
Do you mean all of the branches or just the `dchar` one? If you mean the `dchar` one, using `startsWith` makes it work with empty ranges, though it could check for empty and then check front. If you mean all of them, I don't know if that works. Does `\uFEFF` encode to the BOM for each encoding? I was going off of what wikipedia and std.encoding said the bytes were.

Reply:
> Does \uFEFF encode to the BOM for each encoding? 

Yes, that's how it's defined. From Wikipedia: "The byte order mark (BOM) is a Unicode character, U+FEFF byte order mark (BOM), [...]"

Reply:
> Do you mean all of the branches

Yes, all branches.

> though it could check for empty

Right.

Reply:
It will if you let auto-decoding do its thing... *shudder*

Reply:
No, it won't, because that only works for strings. I mean actual ranges of `char` or `wchar`. I can use `decodeFront` explicitly though.